File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/MyEnum.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.somelib_h;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.lang.foreign.SegmentAllocator;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;

public enum MyEnum {
    A,
    B,
    C,
    D,
    E,
    F,
    ;

    static MyEnum fromInt(int i) {
        switch (i) {
            case -2 -> {
                return MyEnum.A;
            }
            case -1 -> {
                return MyEnum.B;
            }
            case 0 -> {
                return MyEnum.C;
            }
            case 1 -> {
                return MyEnum.D;
            }
            case 2 -> {
                return MyEnum.E;
            }
            case 3 -> {
                return MyEnum.F;
            }
            
        }
        throw new RuntimeException("Unexpected int for MyEnum:" + i);
    }

    int toInt() {
        switch (this) {
            case A -> {
                return -2;
            }
            case B -> {
                return -1;
            }
            case C -> {
                return 0;
            }
            case D -> {
                return 1;
            }
            case E -> {
                return 2;
            }
            case F -> {
                return 3;
            }
            
        }
        throw new RuntimeException("Unexpected variant for MyEnum:" + this);
    }

    
    public static MyEnum getA() {
        
        var nativeVal = somelib_h.MyEnum_get_a();
        
        var returnVal = MyEnum.fromInt(nativeVal);
        return returnVal;
                
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/MyEnum.kt
package dev.diplomattest.somelib

import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface MyEnumLib: Library {
    fun MyEnum_into_value(inner: Int): Byte
    fun MyEnum_get_a(): Int
}
enum class MyEnum(val inner: Int) {
    A(-2),
    B(-1),
    C(0),
    D(1),
    E(2),
    F(3);

    fun toNative(): Int {
        return this.inner
    }


    companion object {
        internal val libClass: Class<MyEnumLib> = MyEnumLib::class.java
        internal val lib: MyEnumLib = Native.load("somelib", libClass)
        fun fromNative(native: Int): MyEnum {
            return when (native) {
                -2 -> A
                -1 -> B
                0 -> C
                1 -> D
                2 -> E
                3 -> F
                else -> throw RuntimeException("Failed to find variant ${native} of type MyEnum")
            }
        }

        fun default(): MyEnum {
            return A
        }
        @JvmStatic
        
        fun getA(): MyEnum {
            
            val returnVal = lib.MyEnum_get_a();
            return (MyEnum.fromNative(returnVal))
        }
    }
    
    fun intoValue(): Byte {
        
        val returnVal = lib.MyEnum_into_value(this.toNative());
        return (returnVal)
    }
}